### PR TITLE
refactor: make ctor/dtor function of singleton classes to be private

### PR DESCRIPTION
**Pull request by levy5307**
make ctor functions of singleton to be private, to avoid creating more than one object for singleton classes

https://github.com/apache/incubator-pegasus/issues/887

### CODE REVIEW

**Review by hycdong — APPROVED** [review]

**Comment by levy5307** [comment]
> Should we also make the deconstructor as private?

I think it doesn't make sense. No one would call the destructor explicitly, and what I want to do is to prevent create a sington outside of instance() function.

**Comment by levy5307** [comment]
> Should we also make the deconstructor as private?

It looks great to me : )

File: include/dsn/utility/singleton.h [hunk 1]
```diff
@@ -17,19 +17,17 @@
...
 
 template <typename T>
-class singleton
+class singleton : private boost::noncopyable
 {
 public:
     singleton() = default;
 
-    // disallow copy and assign
-    singleton(const singleton &) = delete;
```
Comment (levy5307) [inline include/dsn/utility/singleton.h:30]: It is more clear to use `boost::noncopyable`

**Review by acelyc111 — APPROVED** [review]

**Comment by Smityz** [comment]
Maybe we should add some comments in `include/dsn/utility/singleton.h` to tell others how to use it, like setting its deconstructor as private, and tell them why. It's more friendly for c++ beginners

**Comment by levy5307** [comment]
> Maybe we should add some comments in `include/dsn/utility/singleton.h` to tell others how to use it, like setting its deconstructor as private, and tell them why. It's more friendly for c++ beginners

Done

**Review by Smityz — APPROVED** [review]

**Review by hycdong — APPROVED** [review]

**Pull request merged**